generate_perror/main.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"regexp"
+	"sort"
+	"strconv"
+)
+
+const (
+	fileHeader = `// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Generated by generate_perror/main.go
+package main
+
+var MysqlErrNameToNum = map[string]int{
+`
+)
+
+var (
+	tidbCodePath string
+)
+
+func init() {
+	flag.StringVar(&tidbCodePath, "path", "../tidb", "Path to TiDB source code root directory.")
+}
+
+func checkNewErr(errCode string, i int, nameToNum map[string]int) {
+	if v, ok := nameToNum[errCode]; ok {
+		if i != v {
+			if errCode != "handler" {
+				// Ignore the HA_ERR codes, which are all 'handler'
+				log.Printf("Duplicate error errCode %s (%d != %d)", errCode, i, v)
+			}
+		}
+	} else {
+		nameToNum[errCode] = i
+	}
+}
+
+func scanErrCodeFile(fileName string, nameToNum map[string]int) {
+	f, err := os.Open(fileName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+	s := bufio.NewScanner(f)
+	r := regexp.MustCompile(`^\s+(\w+)*\s+=\s+(\d+)$`)
+	for s.Scan() {
+		m := r.FindStringSubmatch(s.Text())
+		if m != nil && len(m) == 3 && m[1] != "" && m[2] != "" {
+			i, err := strconv.Atoi(m[2])
+			if err != nil {
+				log.Fatal(err)
+			}
+			checkNewErr(m[1], i, nameToNum)
+		}
+	}
+	if s.Err() != nil {
+		log.Fatal(s.Err())
+	}
+}
+
+type ErrorCodeAndName struct {
+	Code int
+	Name string
+}
+
+func main() {
+	NameToNum := make(map[string]int)
+
+	// First extract the known error names => numbers from TiDB errno module
+
+	// Second extract the known error names => numbers from TiDB parser/mysql module
+
+	// Last use the perror program to extract error names from 1..20000 from MySQL
+
+	flag.Parse()
+
+	scanErrCodeFile(tidbCodePath+"/pkg/errno/errcode.go", NameToNum)
```

Comment:
```suggestion
	scanErrCodeFile(filepath.Join(tidbCodePath, "/pkg/errno/errcode.go"), NameToNum)
```

Doesn't matter much in this case. Feel free to ignore.